### DISCUSSION TITLE
Border styles not applying in custom layers

### DISCUSSION THREAD

**Discussion by TTransmit** [tailwindlabs/tailwindcss#9811] [Help] [answered]
**What version of Tailwind CSS are you using?**

v3.2.2

**What build tool (or framework if it abstracts the build tool) are you using?**

Next.js 12.3.1, also see the issue with vanilla JS

**What version of Node.js are you using?**

v16.16.0

**What browser are you using?**

Chrome, Safari

**What operating system are you using?**

macOS

**Reproduction URL**
https://play.tailwindcss.com/aaZHDFRQKX?file=css

**Describe your issue**

Borders don't apply from custom layers while other styles do apply. I'm not sure if custom layers are supported but it was an unexpected behaviour. Borders work from `@layer base` and `@layer components` but not from `@layer customlayer`. 

My main use case for a custom layer was as a way to differentiate between styles for different features - visually in the code. I'm guessing I should be using something like postcss-import and have different files for a large project using tailwind. 

```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {

    .border-shows-with-this {
        @apply h-48 w-48 bg-red-600 m-4 border-green-600 border-4;
    }
}

@layer components {

    .border-also-shows-with-this {
        @apply h-48 w-48 bg-red-600 m-4 border-green-600 border-4;
    }
}

@layer different-layer {

    .border-missing-with-this {
        @apply h-48 w-48 bg-red-600 m-4 border-green-600 border-4;
    }
}
```

**Comment by thecrypticace** [comment, marked as answer]
So you're using custom layers but those are a specific feature of CSS itself called cascade layers. Our three `@layer`s (base, components, and utilities) predate that CSS feature. I'm like 99% sure this is happening because our preflight styles are not in a layer and the order of importance for cascade puts explicit layers _before_ un-layered styles.

You can see some more information on this: https://css-tricks.com/css-cascade-layers/#aa-where-do-layers-fit-in-the-cascade

edit: To provide more info:

Our preflight gives every element a zero width border by default. Since un-layered styles are considered more important in the spec this means that `border-4` inside an `@apply` won't do anything because the layer declaration of `border-width: 0` from preflight overrides it.

It would indeed suggest using `postcss-import` to organize your styles if that is something you need to do.

**Reply by TTransmit** [reply]
Thanks, that's helpful. I hadn't heard of CSS cascade layers before: looks like they are pretty new and I'm mainly a mobile developer. It's unfortunate that they use the same at rule name. It's easy enough to work around. Is it worth mentioning in https://tailwindcss.com/docs/adding-custom-styles#using-css-and-layer as a potential gotcha?

**Reply by thecrypticace** [reply]
Yeah, perhaps. I'll discuss this internally and see what the consensus is. Thanks!
